[4.3] Message after a login / logout frontend

sandewt opened the pull request:
Pull Request for Issue # .

New feature, see #30767

### Summary of Changes

There is a message after a successfully login and logout in the frontend.
### Messages
![new_feature_messages_3](https://user-images.githubusercontent.com/9271775/94781992-e8444f00-03ca-11eb-947c-fc4c1bdddfd5.png)
![new_feature_messages_2](https://user-images.githubusercontent.com/9271775/94781998-eaa6a900-03ca-11eb-9ec6-03ace2eb2a22.png)
[EDIT] Removed _successfully_ from the text in message.

### Users: Options
![new_feature_messages_4](https://user-images.githubusercontent.com/9271775/94805383-97455280-03ec-11eb-8eeb-f63ffe7fdc51.png)
[EDIT] Removed

### Testing Instructions

Backend: Go to Users > Manage > Options
(Frontend: Is not applicable.)

Set the options:
- Frontend Login Message (= Yes, default = No)
- Frontend Logout Message (= Yes, default = No)

Create a menu link:
- Menu Login form
- Menu Logout form

1) Log in and out, in case of a:
- Login form (default)
- Menu Login form
- Menu Logout form

2) Code review (developers);

- Check if the code is correct and secure.

### Actual result BEFORE applying this Pull Request

There is NOT showing a message after a suscessfully login or logout.

### Expected result AFTER applying this Pull Request

- There is showing a message after a suscessfully login. 
- There is showing a message after a suscessfully logout.

### Documentation Changes Required

Yes.

sandewt committed:
Message after a login / logout

sandewt committed:
Message after login / logout

sandewt committed:
Message after login / logout

sandewt committed:
Message after login / logout

HLeithner commented:
Please rebase on 4.1-dev

SharkyKZ commented on administrator/components/com_users/config.xml:
```diff
@@ -122,6 +122,30 @@
 		</field>
+		
+		<field
+			name="frontend_login"
```
Can you change this to something less misleading? E.g. `frontend_login_message`.

sandewt replied:
Done

sandewt committed:
Some changes

name="frontend_login_message" =>  name="frontend_login"
filter="boolean " => filter="integer"

sandewt committed:
Some changes

frontend_login => frontend_login_message

sandewt committed:
Some changes

Whitespaces

sandewt commented:
>Please rebase on 4.1-dev

It's just a small extension, a new feature might be a bit exaggerated.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

Quy commented on language/en-GB/com_users.ini:
```diff
@@ -31,6 +31,8 @@ COM_USERS_FIELD_RESET_CONFIRM_USERNAME_LABEL="Username"
 COM_USERS_FIELD_RESET_PASSWORD1_LABEL="Password"
 COM_USERS_FIELD_RESET_PASSWORD1_MESSAGE="The passwords you entered do not match. Please enter your desired password in the password field and confirm your entry by entering it in the confirm password field."
 COM_USERS_FIELD_RESET_PASSWORD2_LABEL="Confirm Password"
+COM_USERS_FRONTEND_LOGIN_SUCCESS="You have been successfully logged in."
```
Remove `successfully`. See https://github.com/joomla/joomla-cms/pull/14607

sandewt replied:
Done.

maikol-ortigueira commented:
I have tested this item :white_check_mark: successfully on 963ad7deca4ef06a722b95009dac832fa42b7a98<br /><br />I have tested the patch and it works perfectly<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

pabloarias commented:
I have tested this item :white_check_mark: successfully on 963ad7deca4ef06a722b95009dac832fa42b7a98<br /><br />Tested successfully on Joomla 4.0.0-beta5-dev and PHP 7.3.22-1

Thank you!

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt committed:
Removed successfully

alikon commented:
i'll not set this PR as RTC despite  2 good tests cause of https://github.com/joomla/joomla-cms/pull/30834#issuecomment-701965397
i'll let the final decision to RL's

brianteeman commented:
We might as well all go home if this wont get merged

sandewt commented:
>Some checks were not successful: continuous-integration/drone/pr — Build is failing

What does this mean? Action?

HLeithner commented:
> We might as well all go home if this wont get merged

We have 27 release blocker and most of the time I only see some new stuff if no one fixes the release blocker and only add new fancy stuff then you are right and we can go home.
https://github.com/joomla/joomla-cms/issues?q=is%3Aopen+is%3Aissue+label%3A"Release+Blocker"

sandewt commented:
> We have 27 release blocker and most of the time I only see some new stuff if no one fixes the release blocker and only add new fancy stuff then you are right and we can go home.

I will close this PR and make a new PR based on [4.1].

HLeithner commented:
You don't need to, I changed your target branch to 4.1-dev

sandewt commented:
> You don't need to, I changed your target branch to 4.1-dev

I see it has already happened. :smiley:

Thanks.

sandewt commented:
>Please rebase on 4.1-dev

Please test in Joomla [4.1].

@maikol-ortigueira / @pabloarias<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

tushar33 commented:
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

particthistle commented:
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />Nice addition @sandewt. Too many times have I not realised I was logged in or out when wanting to be the reverse after clicking the button in J3.

Testing scenarios all worked.

Please add Documentation Required tag so that it can be noted to go back and review relevant help and JDocs.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

zero-24 reviewed (approved):
LTGM @bembelimen please take a look here too :+1:

bembelimen commented:
I really like the feature (good idea) but I'm not sure about the parameter tbh. Is it needed at all? Perhaps we should (in a future version) go for a message management extension where all this notifications can be managed on one point?
Any thoughts?

zero-24 commented:
What kind of `message management extension` are you looking for? Shouldt a module / plugin / component take care of its own messages?

brianteeman commented:
Something like the mail templates?

bembelimen commented:
Yes exactly.

snehaM26 commented:
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />I have tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

ashvini77 commented:
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />Its showing a message on login and logout on frontend.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt commented:
Thanks all.

bembelimen commented:
Again: I would merge it, but I think the parameter are not necessary. We should avoid to add parameter for each thing (yeah I know we did it in the past very often...). If someone want to disable the messages, he/she/it can just make a language override to an empty string.

Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

Could you adjust the PR for this?

sandewt commented:
>Again: I would merge it, I would merge it, but I think the parameter are not necessary.

Sorry, I read about this. Do you mean both parameters?
For now I would like to leave it that way, until there is a centrally regulated message system. Or if several want it that way.

>Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

You mean:

Change this code:
```
// Show a message when a user is logged in.
if ($login === 1 && Factory::getUser()->get('id') > 0)
```
to:
```
// Show a message when a user is logged in.
if (ComponentHelper::getParams('com_users')->get('frontend_login_message', 0))
{
```

bembelimen commented:
> Do you mean both parameters?

Yes both parameters.

sandewt commented:
>If someone want to disable the messages, he/she/it can just make a language override to an empty string.

I have the following notes:

1) Not everyone likes a login / outlog message and can get irritated by it. Hence the default parameters are hide.
2) Especially novice users of Joomla are often unfamiliar with overrides.

sandewt committed:
Delete check if user is really logged in

Delete check if user is really logged in

sandewt commented:
> Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

Done. Deleted check if user is logged in<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

Tejashrimajage commented:
I have tested this item :white_check_mark: successfully on ce1bf617029d4b3ffd244a0f223d74ec78c1f78e<br /><br />It's working fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

vaibhavsTekdi commented:
I have tested this item :white_check_mark: successfully on ce1bf617029d4b3ffd244a0f223d74ec78c1f78e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

HLeithner commented:
@sandewt please remove the parameters they are unneeded and using the override component should be possible with documentation thx

richard67 commented:
Needs some change as stated above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt committed:
Merge branch '4.1-dev' into patch-1

sandewt commented:
> Removed: ? ? ? ? 

See [https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375](https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375 "https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt committed:
Delete both parameters

Delete both parameters

sandewt committed:
Update config.xml

sandewt committed:
Delete two parameters languages strings

Delete two parameters languages strings

richard67 commented:
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt committed:
Delete code two parameters options

Delete code two parameters options

sandewt committed:
Delete code two parameters options

Delete code two parameters options

sandewt commented:
> Back to pending.

@HLeithner and @bembelimen 

Deleted the two parameters.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt commented:
Please test.

- Install the patch.
- Login and logout and see the messages.

That's all, the options have been removed.

Quiviro commented:
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />The messages are shown after login/logout<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

BertaOctech commented:
I have tested this item :red_circle: unsuccessfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />I am sorry but after applyingg the patch I do not see the messages

I am testing on 4.1.0-rc1

Some system info:
------------
 Database Type 	mysql
Database Version 	10.4.19-MariaDB
Database Collation 	utf8mb4_general_ci
Database Connection Collation 	utf8mb4_general_ci
Database Connection Encryption 	None
Database Server Supports Connection Encryption 	No
PHP Version 	7.4.20
Web Server 	Apache/2.4.48 (Unix) OpenSSL/1.1.1k PHP/7.4.20 mod_perl/2.0.11 Perl/v5.32.1
WebServer to PHP Interface 	apache2handler
Joomla! Version 	Joomla! 4.1.0-rc1 Release Candidate [ Kuamini ] 18-January-2022 18:00 GMT
User Agent 	Mozilla/5.0 (X11; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0 
---------------<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt commented:
>I am sorry but after applyingg the patch I do not see the messages
>I am testing on 4.1.0-rc1

I have no explanation for this.

Can you test with Joomla! **4.1.0-dev** Development [ Kuamini ]

BertaOctech commented:
Hi again.

I am sorry but could not test the patch successfully.

I took a fresh install from https://developer.joomla.org/nightly-builds.html
I configured the patch tester but for some reason the oldest pull id available to me was #36134

I then tried to apply the patch manually but I did not get the expected result.

Best regards,

Berta

pabloarias commented:
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />Hello:

My currently installed Joomla! version is ‎4.1.0-rc3-dev. I'm using PHP 8.0.15.

I've created a menu item to login and another to logout.

Before apply path: it doesn't show messages after login or logout.

After applying the path: it shows a message in both cases. After login and logout. Via module or menu items.

Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

infograf768 commented:
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

infograf768 commented:
rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

sandewt commented:
Thanks all. :smiley:

sandewt committed:
Merge branch '4.1-dev' into patch-1

sandewt commented:
Oops, I see that this feature is not yet added to J4.1.
So, I have merged this branche.

sandewt committed:
Merge branch '4.1-dev' into patch-1

richard67 commented:
@sandewt No idea. I am not the only maintainer.

laoneo committed:
Merge branch '4.1-dev' into patch-1

laoneo commented:
This change should go into 4.2. I changed the base branch, but now there are some commits added to your pr which will disappear when 4.1 got fully merged up into 4.2.

richard67 commented:
> > Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.
> 
> Done. Deleted check if user is logged in

@sandewt I see the check has been deleted in the login method but not in the logout method. Does that have a reason, or was that just missed?

laoneo committed:
Merge branch '4.2-dev' into patch-1

sandewt commented:
@richard67 A good comment.
I don't know if there is a check somewhere in Joomla if the user is really logged out at that time. I'll go for safety then.

bembelimen committed:
Merge branch '4.2-dev' into patch-1

bembelimen commented:
> @richard67 A good comment. I don't know if there is a check somewhere in Joomla if the user is really logged out at that time. I'll go for safety then.

No need to be "safe" here, as if the logout failed, it would have been caught before. Although the check would be: ```if ($app->getidentity()->guest)``` (we shouldn't introduce deprecated code) but as mentioned it's not needed.

laoneo commented on components/com_users/src/Controller/UserController.php:
```diff
@@ -228,6 +232,12 @@ public function logout()
 		}
 
+		// Show a message when a user is logged out.
+		if (Factory::getUser()->get('id') === 0)
```
```suggestion
		if ($app->getIdentity()->guest)
```

laoneo replied:
```suggestion
		if ($this->app->getIdentity()->guest)
```

sandewt committed:
deprecated code

Changed deprecated code

sandewt committed:
Better code

`$this->app`
becomes
`$app`

richard67 commented:
@bembelimen @laoneo Do you think we need new tests after the last 2 commits?

laoneo commented:
I think one test for confirmation is enough

sandewt commented:
>I think one test for confirmation is enough

@laoneo I think not
```
echo Factory::getUser()->get('guest') . '<br>'; // let's see 1
echo Factory::getUser()->get('id') . '<br>'; // let's see  0
echo $app->getIdentity()->guest  . '<br>'; // let's see 0
echo $app->getIdentity()->id; // let's NOT see 0
```

sandewt commented:
I think this PR is not successful for the above reasons.

sandewt commented:
In summary:

When using this method, the user IS logged out

`if (Factory::getUser()->get('guest'))`

And when using the following method, the user is NOT logged out

`if ($app->getIdentity()->guest)`

Is completely unclear to me.

bembelimen commented:
Just make it easy: remove the if... messages are enough

sandewt commented:
>Just make it easy: remove the if... messages are enough

The moment the message 'You have been logged out.' is generated, the user is actually still logged in.
This is equivalent to flying the flag before finishing.
Of course, somewhere in the logout process, things will be checked, but still.

That's why I'm hesitating to drop the if statement. Give me a good reason to remove it.

roland-d commented:
@sandewt 
> The moment the message 'You have been logged out.' is generated, the user is actually still logged in.

That is not the case right, the logout is called before that on line 167
https://github.com/joomla/joomla-cms/pull/30834/files#diff-d3df0595ae1fa944f7270d2b86a20ae9358c93d7206c2191da23aebd8ce585a0R167

Do you want to figure out the discrepancies between getting the guest status?

Your comment here https://github.com/joomla/joomla-cms/pull/30834#issuecomment-1067956161 may need some further investigation in a separate issue.

joomla-bot committed:
Merge tag 'psr12anchor' into psr12/merge/30834

joomla-bot committed:
Phase 1 convert BRANCH to PSR-12

joomla-bot committed:
Phase 2 convert BRANCH to PSR-12

joomla-bot committed:
Merge tag 'psr12final' into psr12/merge/30834

richard67 commented:
Back to pending (RTC was added back by the bot by accident).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

laoneo committed:
Merge branch '4.3-dev' into patch-1

drmenzelit committed:
Merge branch '4.3-dev' into patch-1

drmenzelit committed:
Merge branch '4.3-dev' into patch-1

drmenzelit commented:
Thanks

sandewt commented:
> Your comment here https://github.com/joomla/joomla-cms/pull/30834#issuecomment-1067956161 may need some further investigation in a separate issue.

@drmenzelit 

**Something went wrong here. This PR is merged to joomla 4.3. That shouldn't have happened.
Because this PR is not working.**

sandewt commented:
Removing the if statement makes this PR work. The question is whether that is desirable. See the discussion above.

```
// Show a message when a user is logged out.
if ($app->getIdentity()->guest) {
    $app->enqueueMessage(Text::_('COM_USERS_FRONTEND_LOGOUT_SUCCESS'), 'message');
}
```

```
// Show a message when a user is logged out.
$app->enqueueMessage(Text::_('COM_USERS_FRONTEND_LOGOUT_SUCCESS'), 'message');
```

bembelimen commented:
Please test #39048

richard67 commented:
@sandewt Please test #39048 . Thanks in advance.

sandewt commented:
@richard67 

I'd like to do that, but then I'm actually testing my own code proposal. I wonder if that's desirable :wink:
